File: exercisesC/coins/coinsCredit.c
#include <stdio.h>
#include <math.h>

int validCreditCard(int totalDigits,int digit[]);
void printCardFlag(int digit[],int totalDigits);

int main(){

    long ccNumber = 0;

    printf("Type the number of credit card: ");
    scanf("%ld",&ccNumber);


    long aux = ccNumber;
    int totalDigits = 0;

    while(aux != 0){
        aux = aux / 10;                         //count total of digits
        totalDigits += 1;
    }

    int digit[20];
    long divisor = 1;
    for(int i = 0 ; i<totalDigits ; i++){
        digit[i] = (ccNumber / divisor) % 10;  //create a vector with each card number in each possition
        divisor *= 10;
    }

    printCardFlag(digit,totalDigits);

    if(validCreditCard(totalDigits,digit) == 1){
        printf("Valid credit card");
    }else{
        printf("inValid credit card");
    }

    printf("\n");

    return 0;
}

int validCreditCard(int totalDigits,int digit[]){
    int sum = 0;
    for(int i = 1 ; i<totalDigits ; i = i+2){
        int temp = (digit[i]*2);
        if(temp > 9){
            sum += (temp/10);
            sum += (temp%10);
        }else{
            sum += temp;
        }
        sum += digit[i-1];
    }

    int lastDigit = sum%10;

    if(lastDigit == 0)return 1;
    return 0;
}

void printCardFlag(int digit[],int totalDigits){

    int ccFlag=0;
    ccFlag +=(digit[totalDigits-1] * 10);
    ccFlag += digit[totalDigits-2];

    if(ccFlag == 55 || ccFlag == 54 || ccFlag == 53 || ccFlag == 52 ||ccFlag == 51){
        printf("Mastercard \n");
    }else if(ccFlag == 34 || ccFlag == 37){
        printf("American Express\n");
    }else if(ccFlag/10 == 4){
        printf("Visa\n");
    }else{
        printf("Invalid card flag\n");
    }
}

File: exercisesPython/coins/coinsCredit.py
def main():
    number = input("Number:")

    firstDigits = int(number[0]) * 10 + int(number[1])

    if(len(number) == 15 and (firstDigits == 34 or firstDigits == 37)):
        cardType = "American Express"
        valid = luhn(number)

    elif(len(number) == 13 and int(number[0]) == 4):
        cardType = "Visa"
        valid = luhn(number)

    elif(len(number) == 16):
        if(int(number[0]) == 4):
            cardType = "Visa"
            valid = luhn(number)
        elif(firstDigits in (51,52,53,54,55)):
            cardType = "Mastercard"
            valid = luhn(number)
        else:
            valid = False

    else:
        valid = False

    if(valid == False):
        cardType = "Invalid"

    print(cardType)



def luhn(card):
    tmp = 0
    sum = 0
    for i in range(len(card)-1 , 0 ,2):
        tmp = int(card[i]) * 2
        if(tmp >= 10):
            sum += tmp // 10
            sum += tmp % 10
        else:
            sum += tmp

    for i in range(len(card), 0, 2):
        sum += int(card[i])

    print(sum)
    if(sum % 10 == 0):
        return True
    else:
        return False

main()
